pyomo/gdp/plugins/hull.py

```diff
@@ -462,35 +489,86 @@ def _transform_disjunctionData(self, obj, index, transBlock=None):
                 else:
                     # It's not local to this Disjunct
-                    varSet.append(var)
+                    varSet[disjuncts[0]].append(var)
+                    varsToDisaggregate.append(var)
             else:
                 # We don't even have have any local vars for this Disjunct.
-                varSet.append(var)
+                varSet[disjuncts[0]].append(var)
+                varsToDisaggregate.append(var)
 
         # Now that we know who we need to disaggregate, we will do it
         # while we also transform the disjuncts.
         or_expr = 0
         for disjunct in obj.disjuncts:
-            or_expr += disjunct.binary_indicator_var
-            self._transform_disjunct(disjunct, transBlock, varSet,
+            or_expr += disjunct.indicator_var.get_associated_binary()
+            self._transform_disjunct(disjunct, transBlock, varSet[disjunct],
                                      localVars.get(disjunct, []))
         orConstraint.add(index, (or_expr, 1))
         # map the DisjunctionData to its XOR constraint to mark it as
         # transformed
         obj._algebraic_constraint = weakref_ref(orConstraint[index])
 
         # add the reaggregation constraints
-        for i, var in enumerate(varSet):
-            disaggregatedExpr = 0
-            for disjunct in obj.disjuncts:
+        for i, var in enumerate(varsToDisaggregate):
+            # There are two cases here: Either the var appeared in every
+            # disjunct in the disjunction, or it didn't. If it did, there's
+            # nothing special to do: All of the disaggregated variables have
+            # been created, and we can just proceed and make this constraint. If
+            # it didn't, we need one more disaggregated variable, correctly
+            # defined. And then we can make the constraint.
+            if len(disjunctsVarAppearsIn[var]) < len(obj.disjuncts):
+                # create one more disaggregated var
+                idx = len(disaggregatedVars)
+                disaggregated_var = disaggregatedVars[idx]
+                lb = var.lb
+                ub = var.ub
+                if lb > 0:
+                    disaggregated_var.setlb(0)
+                else:
+                    disaggregated_var.setlb(lb)
+                if ub < 0:
+                    disaggregated_var.setub(0)
+                else:
+                    disaggregated_var.setub(ub)
+
+                # and we need the bounds constraints too
+                var_free = 1 - sum(disj.indicator_var.get_associated_binary()
+                                   for disj in disjunctsVarAppearsIn[var])
+                disaggregated_var_bounds[idx, 'lb'] = lb*var_free <= \
+                                                      disaggregated_var
+                disaggregated_var_bounds[idx, 'ub'] = disaggregated_var <= \
+                                                      ub*var_free
```

Comment:
This is basically repeated elsewhere.  Can we combine this into a single helper function?

Reply:
OK this is done. It was a bigger change because it caught a couple bugs: 

1. Lines 535-540 above are wrong because the bounds constraints should only get added (respectively) if lb and ub are not `None` and nonzero. (That's what changed the hull test and the FME tests.)
2. In the check to see if a Disjunction was nested (if we should declare its disaggregated variables local), I was checking the parent Block of the Disjunct, not the Disjunction. (This is the wrong check, see the `makeNestedDisjunctions_FlatDisjuncts` model.) I moved this into a helper function as well, as it is needed for both the normal disaggregated variables and the ones representing multiple Disjuncts.

This broke a create_using test though, so we need to fix that before we can merge.